jscomp/others/js_promise2.res

```diff
@@ -0,0 +1,18 @@
+include Js_promise
+
+/** Type-safe t-first then */
+let then: (promise<'a>, 'a => promise<'b>) => promise<'b> = %raw(`
+  function(p, cont) {
+    Promise.resolve(p).then(cont)
```

Comment:
Why do you wrap a promise with `Promise.resolve`?

Reply:
Because promises are flattened at runtime and where you expect to have, say `promise<int>` because you looked inside `promise<promise<int>>` what you actually get is an `int`. When that happens, the `int` is converted back to the `promise<int>` you were expecting, so you can call `.then` on it.

Reply:
> Because promises are flattened at runtime and where you expect to have, say promise<int> because you looked inside promise<promise<int>> what you actually get is an int.

Once it became a promise it will stay a promise. So if we have `promise<promise<int>>` in runtime, it'll be flattened to `promise<int>` (not `int`). So we can call `.then` without additional `Promise.resolve`.

At least it works this way in JavaScript. Maybe we're talking about different things.

Reply:
Not easy to explain in few words. But you can check the example: https://github.com/rescript-lang/rescript-compiler/pull/5709/files#diff-dfcc91a2765ccd42f814d348224f3bc20b343386851e14fd99132000aa56fd78R13

It is flattened to `promise<int>`, but when you "look inside" i.e. await it, you get `int`. (Or you could look inside by doing another `.then`).


Another way to answer: in JS `await` is happy to receive a value such as a number, it does not need a promise. This was done intentionally. This change does the same for `then`: it makes `then` happy to receive a value such as a number.

Reply:
Ok, I think I need to go to the PR https://github.com/rescript-lang/rescript-compiler/issues/5707

Reply:
>For example, if xxx has the type promise<promise<promise<int>>>, the Promise2 will also stop working correctly.

Feel free to write the example and if you observe a runtime crash, post back here.

Reply:
```
let nestedPromise = async (xxx: promise<promise<promise<int>>>) => {
  let xx = await xxx
  //  ^ It's `promise<promise<int>>`

  let _ = xx->Js.Promise2.then(x => Js.log2("Promise2.then", x) |> Js.Promise.resolve)
  //                                              ^^^ It's `promise<int>`
  // The code compiles, but x has the wrong type
}
```

Reply:
What I suggest is:
```
let nestedPromise = async (xxx: promise<int>) => {
  let xx = await xxx
  Js.log2("Promise2.then", xx)
}

module Promise2 = {
  external flatten: promise<promise<'a>> => promise<'a> = "%identity"
}

let someFnThatReturnsNestedPromises: unit => promise<promise<promise<int>>> = () => %raw("Promise.resolve(123)")

someFnThatReturnsNestedPromises()
  ->Promise2.flatten
  ->Promise2.flatten
  ->nestedPromise
  ->ignore
```
This way, we solve the problem immediately as it's raised. No twisting of the `promise` type meaning. Also, the `nestedPromise` code became much cleaner.

Reply:
This is the same example. Something of type `promise<t>` contains an int. We know that and it was already true in the previous example.
There's no claim that this is nice, but avoiding this is harder than it seems.

Btw, the same crashing example can be written using only promises, but without `async`.
Interestingly, the crashing example cannot be written using only `async` without also writing direct promise primitives.

About trying to flatten the promise type. You could write the same example with `xxx: promise<t>` and a few abstract functions on `t` without ever knowing what `t` is. The body of the function would never show `promise<promise<...>` nor would any code calling the function. The only way to discover that a nested promise is used would be, every time the function is called, to type check the body again, and see if for the specific instance of `t`, the type could have been flattened, to discover that the program could crash.

Reply:
> What I suggest is:
> 
> ```
> let nestedPromise = async (xxx: promise<int>) => {
>   let xx = await xxx
>   Js.log2("Promise2.then", xx)
> }
> 
> module Promise2 = {
>   external flatten: promise<promise<'a>> => promise<'a> = "%identity"
> }
> 
> let someFnThatReturnsNestedPromises: unit => promise<promise<promise<int>>> = () => %raw("Promise.resolve(123)")
> 
> someFnThatReturnsNestedPromises()
>   ->Promise2.flatten
>   ->Promise2.flatten
>   ->nestedPromise
>   ->ignore
> ```
> 
> This way, we solve the problem immediately as it's raised. No twisting of the `promise` type meaning. Also, the `nestedPromise` code became much cleaner.

I think there's some confusion on what the problem is.
The problem is that you can inadvertently write some code that uses nested promises and crash.
The nested promise type can easily be in the middle of an expression, and the user might never see it written down explicitly.
If you already know there are nested promises, then clearly there's no problem and you can flatten them explicitly.
The problem this is trying to solve is when you don't realise you're using nested promises.

Reply:
I consider crashing a healthy thing when it happens during development. I think it's where we derail in our opinions.

Reply:
Especially when using `Js` module. Maybe it's for `Belt` to handle this problem. What do you think?

Reply:
Definitely, wanting those examples to crash is one possible opinion.
What is surprising, at least to me, is that the same example written only in terms of `async/await` does not crash.
This boils down to `await 3` not crashing.

Reply:
> Definitely, wanting those examples to crash is one possible opinion.

It's how it used to work https://github.com/rescript-lang/rescript-compiler/pull/5709/files#diff-dfcc91a2765ccd42f814d348224f3bc20b343386851e14fd99132000aa56fd78R13

Ideally would be nice to have types that automatically flatten eg, `promise<promise<t>>` automatically becomes `promise<t>` on the type system level. The same thing for the `option` type and others. But I can imagine that it's not easy to do.

Well, I don't have anything else to add.  I'm more into leaving things as they are right now. Both approaches have pros and cons. I'm not a decision-maker here.